cinder/scheduler/filters/shard_filter.py

```diff
@@ -142,14 +150,74 @@ def _get_shards(self, project_id):
 
         return self._PROJECT_SHARD_CACHE.get(project_id)
 
-    def backend_passes(self, backend_state, filter_properties):
+    def _is_vmware(self, backend_state):
         # We only need the shard filter for vmware based pools
         if backend_state.vendor_name != 'VMware':
             LOG.info(
-                "Shard Filter ignoring backend %s as it's not vmware based"
-                " driver", backend_state.backend_id)
+                "Shard Filter ignoring backend %s as it's not "
+                "vmware based driver", backend_state.backend_id)
             return True
+        return False
+
+    def filter_all(self, filter_obj_list, filter_properties):
+        backends = self._filter_by_k8s_cluster(filter_obj_list,
+                                               filter_properties)
+
+        return [b for b in backends if
+                self._backend_passes(b, filter_properties)]
+
+    def _filter_by_k8s_cluster(self, backends, filter_properties):
+
+        metadata = filter_properties.get('metadata', {})
+        cluster_name = metadata.get(CSI_CLUSTER_METADATA_KEY)
+
+        spec = filter_properties.get('request_spec', {})
+        vol = spec.get('volume_properties', {})
+        project_id = vol.get('project_id', None)
+
+        if not cluster_name or not project_id:
+            return backends
+
+        query_filters = {
+            "metadata": {CSI_CLUSTER_METADATA_KEY: cluster_name}
+        }
+
+        props = spec.get('resource_properties', {})
+        availability_zone = props.get('availability_zone')
+
+        if availability_zone:
+            query_filters['availability_zone'] = availability_zone
+
+        k8s_volumes = VolumeList.get_all_by_project(
+            cinder_context.get_admin_context(),
+            project_id,
+            filters=query_filters)
```

Comment:
This could be a very expensive call to fetch all the volumes for a project on every scheduler request.

Reply:
As we are not really interested in the volumes themselves, but rather the (hopefully limited) set of hosts,
I would prefer to move the logic as far as possible to the DB, i.e.:

```sql
SELECT SUBSTRING_INDEX(v.host, '@', 1) AS h, count(*) AS n 
FROM volume_metadata AS vm JOIN volumes AS v ON vm.volume_id=v.id
WHERE vm.`key`='cinder.csi.openstack.org/cluster' 
  AND vm.`value`=@cluster_name
  AND v.deleted=0
  AND v.`host` IS NOT NULL
  GROUP BY h;
```

We are missing though an index on `volume_metadata.key` and `volume_metadata.value` though.

This would also eliminate the need for `all_backend_states`.